language/move-prover/bytecode/src/mono_analysis.rs

```diff
@@ -0,0 +1,239 @@
+                        self.analyze_target(target);
+                    }
+                }
+            }
+        }
+        // Now incrementally work included targets until they are done, while self.inst_opt
+        // contains the specific instantiation.
+        while !self.todo_targets.is_empty() {
+            let (fun, inst) = self.todo_targets.pop().unwrap();
+            self.inst_opt = Some(inst);
+            self.analyze_target(
+                self.targets
+                    .get_target(&self.env.get_function(fun), &FunctionVariant::Baseline),
+            );
+            let inst = std::mem::take(&mut self.inst_opt).unwrap();
+            if !inst.is_empty() {
+                // Insert it into final analysis result if not trivial
+                self.info.funs.entry(fun).or_default().insert(inst.clone());
+            }
+            self.done_targets.insert((fun, inst));
+        }
+    }
+
+    fn analyze_target(&mut self, target: FunctionTarget<'_>) {
+        // Analyze function locals and return value types.
+        for idx in 0..target.get_local_count() {
+            self.add_type(target.get_local_type(idx));
+        }
+        for ty in target.get_return_types().iter() {
+            self.add_type(ty);
+        }
+        // Analyze code.
+        if !target.func_env.is_native_or_intrinsic() {
+            for bc in target.get_bytecode() {
+                self.analyze_bytecode(&target, bc);
+            }
+        }
+    }
+
+    fn analyze_bytecode(&mut self, target: &FunctionTarget<'_>, bc: &Bytecode) {
+        use Bytecode::*;
+        use Operation::*;
+        // We only need to analyze function calls, not `pack` or other instructions
+        // because the types those are using are reflected in locals which are analyzed
+        // elsewhere.
+        if let Call(_, _, Function(mid, fid, targs), ..) = bc {
+            if !target.is_opaque() {
+                // This call needs to be inlined, with targs instantiated by self.inst_opt.
+                // Schedule for later processing.
+                let actuals = if let Some(inst) = &self.inst_opt {
+                    targs.iter().map(|ty| ty.instantiate(inst)).collect_vec()
+                } else {
+                    targs.to_owned()
+                };
+                let fun = mid.qualified(*fid);
+                // Only if this call has not been processed yet, queue it for future processing.
+                if !self.done_targets.contains(&(fun, actuals.clone())) {
+                    self.todo_targets.push((mid.qualified(*fid), actuals));
+                }
+            }
+        }
+    }
+
+    // Type Analysis
+    // =============
+
+    fn add_type(&mut self, ty: &Type) {
+        if let Some(inst) = &self.inst_opt {
+            let ty = ty.instantiate(inst);
+            self.add_type2(&ty)
+        } else {
+            self.add_type2(ty)
+        }
+    }
+
+    fn add_type2(&mut self, ty: &Type) {
```

Comment:
I guess the convention is `add_type_` (a suffix underscore) to denote internal function

Reply:
I renamed it, but this convention you mention I only know for escaping keywords, so choose just a prettier name ;-)